docs/docs/server-middleware.md

```diff
@@ -0,0 +1,649 @@
+)(service).orNotFound
+
+```
+```scala mdoc
+loggerService(reverseRequest.withEntity("mood")).flatMap(_.as[Unit])
+  .attempt.void.unsafeRunSync()
+loggerLog
+```
+
+## Advanced
+
+### BodyCache
+
+Consumes and caches a request body so that it can be reused later.
+Usually reading the body twice is unsafe, this middleware ensures the body is always the same,
+at the cost of keeping it in memory.
+
+In this example we use a request body that always produces a different value once read:
+```scala mdoc:silent
+import org.http4s.server.middleware.BodyCache
+
+val bodyCacheService = BodyCache.httpRoutes(service).orNotFound
+
+val randomRequest = Request[IO](Method.GET, uri"/doubleRead")
+  .withEntity(
+    Stream.eval(
+      Random.scalaUtilRandom[IO].flatMap(_.nextInt).map(random => random.toString)
+    )
+  )
+```
+`/doubleRead` reads the body twice, when using the middleware we see that both read values the same:
+```scala mdoc
+service.orNotFound(randomRequest).flatMap(r => r.as[String]).unsafeRunSync()
+bodyCacheService(randomRequest).flatMap(r => r.as[String]).unsafeRunSync()
+```
+
+### BracketRequestResponse
+
+Brackets the handling of the request ensuring an action happens before the service handles
+the request (`acquire`) and another after the response is complete (`release`),
+the result of `acquire` is threaded to the underlying service. 
+It's used to implement `MaxActiveRequests` and `ConcurrentRequests`. 
+See [BracketRequestResponse] for more constructors.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.BracketRequestResponse
+import org.http4s.ContextRoutes
+import cats.effect.Ref
+
+val ref = Ref[IO].of(0).unsafeRunSync()
+
+val bracketMiddleware = BracketRequestResponse.bracketRequestResponseRoutes[IO, Int](
+  acquire = ref.updateAndGet(_ + 1))(
+  release = _ => ref.update(_ - 1)
+)
+
+val bracketService = bracketMiddleware(
+  ContextRoutes.of[Int, IO] {
+    case GET -> Root / "ok" as n => Ok(s"$n")
+  }
+).orNotFound
+
+```
+```scala mdoc
+bracketService(okRequest).flatMap(r => r.as[String]).unsafeRunSync()
+ref.get.unsafeRunSync()
+```
+
+### ChunkAggregator
+
+Consumes and caches a response body so that it can be reused later.
+Usually reading the body twice is unsafe, this middleware ensures the body is always the same,
+at the cost of keeping it in memory.
+
+Similarly to `BodyRequest` in this example we use a response body that always produces a different value:
+```scala mdoc:silent
+import org.http4s.server.middleware.ChunkAggregator
+
+val chunkAggregatorService = ChunkAggregator.httpRoutes(service).orNotFound
+```
+```scala mdoc
+chunkAggregatorService(Request[IO](Method.GET, uri"/random"))
+  .flatMap(r => (r.as[String], r.as[String]).mapN((a, b) => s"$a == $b"))
+  .unsafeRunSync()
+```
+
+### Jsonp
+
+Jsonp is a javascript technique to load json data without using [XMLHttpRequest],
+which bypasses the same-origin security policy implemented in browsers.
```

Comment:
- https://github.com/http4s/http4s/issues/6959

Reply:
I'll follow that issue and make a new PR once things change. Anything else i should do in the meantime?

Reply:
Since the code is currently commented and jsonp is on its way out, should we even bother to keep it?

Alternatively, I guess you can make it an mdoc compile-only block and uncomment the code, or even just an ordinary scala code block that is not checked.

Reply:
If it's all the same to you i'd prefer that the document lists all middleware, feels neater.
I'll mess around with mdoc and add the caveat that jsonp is discouraged.

Reply:
Makes sense to me!